src/vm_pmu.rs

```diff
@@ -368,4 +375,195 @@ impl VmPmuState {
+            sbi::api::pmu::configure_matching_counters(
+                counter_index,
+                counter_mask,
+                config_flags,
+                event_type,
+                event_data,
+            )
+        } else {
+            fn configure_counter_csrs(
+                config_flags: PmuCounterConfigFlags,
+                event_type: PmuEventType,
+                event_data: u64,
+                counter_index: u64,
+            ) {
+                let mut shpmevent = LocalRegisterCopy::<u64, shpmevent::Register>::new(0);
+                // Ignore vuinh / vsinh from the VM and translate uinh
+                if config_flags.is_uinh() {
+                    shpmevent.modify(shpmevent::vuinh.val(1));
+                }
+                if config_flags.is_sinh() {
+                    shpmevent.modify(shpmevent::sinh.val(1));
+                }
+                match event_type {
+                    PmuEventType::Hardware(_) | PmuEventType::Cache(_) => {
+                        shpmevent.modify(shpmevent::event_selector.val(event_type.raw()))
+                    }
+                    PmuEventType::RawEvent => {
+                        shpmevent.modify(shpmevent::event_selector.val(event_data))
+                    }
+                    _ => unreachable!(),
+                }
+                CSR.scounterselect.set(counter_index);
+                CSR.shpmevent.set(shpmevent.get());
+                if config_flags.is_clear_value() {
+                    CSR.shpmcounter.set(0);
+                }
+                if config_flags.is_auto_start() {
+                    CSR.scountinhibit.read_and_clear_bits(1 << counter_index);
+                }
+            }
+
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let index = if !config_flags.is_skip_match() {
+                match event_type {
+                    PmuEventType::Hardware(hardware_event_type) => {
+                        let index = CounterMaskIter::new(counter_index, counter_mask)
+                            .into_iter()
+                            .next()
+                            .map_or(usize::MAX, |i| i);
+                        let is_unconfigured = self
+                            .counter_state
+                            .get(index)
+                            .map_or(false, |counter_state| {
+                                matches!(counter_state, PmuCounterState::NotConfigured)
+                            });
+                        // Fixed counters must use a static mapping pending finalization of event mapping
+                        match hardware_event_type {
+                            PmuHardware::CpuCycles => {
+                                if is_unconfigured && index == PmuHardware::CpuCycles as usize {
+                                    let counter_index = PmuHardware::CpuCycles as u64;
```

Comment:
The event ID is not the same as the counter index. The fixed `cycles` counter is counter 0, while the `cycles` event ID is 1. (They do happen to be the same for `instret`, though.)

Reply:
Aggh...good catch....got them mixed up somehow....fixed.